### ISSUE TITLE
Inserting many to many relationship doesn't seem to work

### ISSUE THREAD

**Issue by zywind** [vuex-orm/vuex-orm#113]
I tried loading data that contains many to many relationships. I found that though the relationship fields are populated, the pivot table isn't. I have code like below:

```js
export default class ClusterExample extends Model {
  static entity = "cluster_example";

  static primaryKey = ["cluster_id", "example_id"];

  static fields() {
    return {
      cluster_id: this.attr(null),
      example_id: this.attr(null)
    };
  }
}

export default class Example extends Model {
  static entity = "examples";

  static fields() {
    return {
      id: this.attr(null),
      text: this.attr(null),
      clusters: this.belongsToMany(
        Cluster,
        ClusterExample,
        "example_id",
        "cluster_id"
      )
    };
  }
}

export default class Cluster extends Model {
  static entity = "clusters";

  static fields() {
    return {
      id: this.attr(null),
      clustering_type: this.attr(null),
      examples: this.belongsToMany(
        Example,
        ClusterExample,
        "cluster_id",
        "example_id"
      )
    };
  }
}
```

And I tried loading the cluster data, which has the examples field. Though the clusters entities are correctly filled, the examples entities' clusters fields are still null, and the pivot table cluster_example is empty.

Is many-to-many relationships not populated automatically?

**Comment by zywind** [comment]
I tried populating the pivot table manually, and it seems to work. However, querying the nested object is horribly slow. I have a query like so:
```
this.$store.getters["entities/clusterings/query"]()
        .with("clusters.examples")
        .all();
```
And it takes about a minute to finish. There are about 2000 records in the pivot table. Is this performance to be expected?

**Comment by kiaking** [comment]
> Is many-to-many relationships not populated automatically?

It will be generated automatically if you save the record that containing the relational records. For example, Vuex ORM has following test case.

```js
class User extends Model {
  static entity = 'users'

  static fields () {
    return {
      id: this.attr(null),
      roles: this.belongsToMany(Role, RoleUser, 'user_id', 'role_id')
    }
  }
}

class Role extends Model {
  static entity = 'roles'

  static fields () {
    return {
      id: this.attr(null),
      users: this.belongsToMany(User, RoleUser, 'role_id', 'user_id')
    }
  }
}

class RoleUser extends Model {
  static entity = 'roleUser'

  static primaryKey = ['role_id', 'user_id']

  static fields () {
    return {
      role_id: this.attr(null),
      user_id: this.attr(null)
    }
  }
}

createApplication('entities', [{ model: User }, { model: Role }, { model: RoleUser }])

const state = {
  name: 'entities',
  users: { data: {} },
  roles: { data: {} },
  roleUser: { data: {} }
}

const data = {
  id: 1,
  roles: [
    { id: 2 },
    { id: 3 }
  ]
}

const expected = {
  name: 'entities',
  users: { data: {
    '1': { $id: 1, id: 1, roles: [2, 3] }
  }},
  roles: { data: {
    '2': { $id: 2, id: 2, users: [] },
    '3': { $id: 3, id: 3, users: [] }
  }},
  roleUser: { data: {
    '1_2': { $id: '1_2', user_id: 1, role_id: 2  },
    '1_3': { $id: '1_3', user_id: 1, role_id: 3 }
  }}
}

// This equals to `dispatch('entities/users/create', { data })`
Repo.create(state, 'users', data)

expect(state).toEqual(expected)
```

In above code, the data object is a user, and the user contains roles within the object. In this case, the pivot table should be populated. However, if save each record separately, then pivot table would not be populated because Vuex ORM has no idea if those two records are connected or not.

Is this what you were doing? If the pivot table is not populated even if you were passing Cluster data with Example data, then that's unexpected :/

> And it takes about a minute to finish. There are about 2000 records in the pivot table. Is this performance to be expected?

To be honest, I haven't tried that amount of record, but taking more than a minute with 2000 records is not expected! 🔥 

I'll try to take benchmark and see where the bottleneck is.

So you have about 2000 ClusterExample records right? May I ask you how many Cluster records are there?

**Comment by zywind** [comment]
Thanks for your quick response!
You're right. I was populating the example data first and then the clusters. I'll populate all the data together now.

I have about 2000 ClusterExample records and 50 cluster records. It would be great if the performance problem can be fixed.

**Comment by zywind** [comment]
I tried inserting parents and children together and the many-to-many relationships were indeed automatically populated. Though I found that automatic relationship insertion seems to only work between parents and children. If I have another level of nested objects, the relationships were lost. So I had to issue several calls to go through every pairs of relationships to insert them all. This is fine, but it would be nice to have this documented.

 The slow query performance still exists with the many-to-many relationship though.

**Comment by kiaking** [comment]
> Though I found that automatic relationship insertion seems to only work between parents and children.

Well, that's unexpected :/ This is another issue I should fix. Thank you for the report!

> The slow query performance still exists with the many-to-many relationship though.

About this problem, I've tried to reproduce, and I did. Yes, it's super slow when you have lots of pivots records (in your case it's ClusterExample). I don't know how long it took since my test timed out. Which means it took so long that I had to quit.

Then I've tweaked few implementations, and I managed to drastically reduce the duration. Now with 50 Cluster, 2000 Example, 2000 Cluster Example and it take about 10ms.

If I make 2000 record each, it takes about 20 to 40ms. If I bump this up to 10,000 records each, it takes about 150ms to 250ms.

I don't know if 250ms for retrieving a total of 30,000 relational records is fast enough, but it's way faster than the current implementation. And I think people shouldn't save that amount of data into JavaScript memory anyway...

So, I'm going to make the fix. There're quite a bit of refactoring has to be done so that it might take a while, but I'll try my best!

**Comment by zywind** [comment]
Wow! Thank you so much! 250 ms for that many records is definitely enough!

**Comment by kiaking** [comment]
The performance issue should've been solved via Vuex ORM 0.23.0 🎉 
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.23.0

Could you try this version to see if everything works for you?

**Comment by zywind** [comment]
Awesome! I will test it ASAP! Thank you so much!

**Comment by zywind** [comment]
Hmm, I'm having some trouble populating the many-to-many association table again. My payload for initializing the store has this structure:
```js
{ // clustering object
    id: "",
    fields: "...",
    clusters: [
        { // cluster object
            id: "",
            fields: "...",
            examples: [
                { // example object
                    id: "",
                    fields: "..."
                }
            ]
        }
    ]
}
```
So it should have all the information needed to populate the cluster_example association table. But while the cluster entities do have their example children filled, the association table is still empty. Furthermore, even if I re-insert from the `clusters` array (so that `examples` is only one level down), the association table is still not filled.

**Comment by zywind** [comment]
I updated the Relations_BelongsToMany test case to below, and it indeed fails:
```js
// line 42
   await store.dispatch('entities/users/create', { data: { id: 1, roles: [{id: 1}, {id: 2}] } })

    const users = store.getters['entities/users/query'].with('roles').all().

    expect(users[0].roles.length).toBe(2)
```

**Comment by zywind** [comment]
I tried manually inserting the association table and then tried querying. It's super fast now. I can't feel any delay!

**Comment by kiaking** [comment]
Great! At least one problem is solved now ; )

I'll look into the automatic pivot table creation issue... Not sure what happened since I didn't change those logic so... Well we'll see!

**Comment by kiaking** [comment]
Thanks to @afsardo, this issue has been solved by #128 🎉 
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.23.4
